Update project documentation and benchmarking (#72)
clflushopt/tpchgen-rs: benchmarks/BENCHMARKING.md

#
-CFLAGS = -g -DDBNAME=\"dss\" -D$(MACHINE) -D$(DATABASE) -D$(WORKLOAD) -DRNG_TEST -D_FILE_OFFSET_BITS=64
+CFLAGS = -g -DDBNAME=\"dss\" -D$(MACHINE) -D$(DATABASE) -D$(WORKLOAD) -DRNG_TEST -D_FILE_OFFSET_BITS=64  -O3
LDFLAGS = -O
# The OBJ,EXE and LIB macros will need to be changed for compilation under
#  Windows NT
```

## `dbgen_docker` docker (`tbl_dbgen_docker.sh`)

Since compiling `dbgen` is non trivial, projects such as DataFusion (TOOD link) often use a pre-built docker image

Example command to generate scale factor 1 dataset into `/tmp`

```sh
docker run -v "/tmp:/data" --rm ghcr.io/scalytics/tpch-docker:main -vf -s 1
```

# Columnar file formats

## parquet tpchdbgen (`parquet_tpchgen.sh`)
Uses the CLI in this repo. Deafults to snappy compression

This uses less than 400MB of peakmemory as measured by top

Example command for SF=10

```shell
tpchgen-cli -s 10 --format=parquet
```

## duckdb duckdb (`duckdb_duckdb.sh.`)

Note duckdb does not create the `tbl` files, but instead creates a database in its own custom format file so we can not compare directly

We use the TPCH data generator as described here https://duckdb.org/docs/stable/extensions/tpch.html

> Note for the larger scale factors, as documneted on https://duckdb.org/docs/stable/extensions/tpch.html, duckdb consumes signifcicant amounts of memoyr

Reviewer: ```suggestion
Note for the larger scale factors, as documented  on https://duckdb.org/docs/stable/extensions/tpch.html, duckdb consumes significant amounts of memory 
```

```sql
INSTALL tpch;
LOAD tpch;

.open test
.timer on
CALL dbgen(sf = 1);
```

duckdb test.duckdb "INSTALL tpch; LOAD tpch; CALL dbgen(sf = 1);"

> Note I did not have a machine with enough memoyr to create scale factor 100

Reviewer: ```suggestion
Note I did not have a machine with enough memory to create scale factor 100
```

> Note I did not have a machine with enough memoyr to create scale factor 100

Reviewer: Potentially link to the statistics released by DuckDB in the TPCH extension page.

Author: Indeed -- good idea

## duckdb duckdb (`parquet_duckdb.sh.`)

Run the above commands for `duckdb` and then export the data to parquet.

note the output parquet file uses SNAPPY compression

```sql
copy customer to 'customer.parquet' (FORMAT parquet);
copy lineitem to 'lineitem.parquet' (FORMAT parquet);
copy nation to 'supplier.parquet' (FORMAT parquet);
copy orders to 'supplier.parquet' (FORMAT parquet);
copy part to 'part.parquet' (FORMAT parquet);
copy partsupp to 'partsupp.parquet' (FORMAT parquet);
copy region to 'region.parquet' (FORMAT parquet);
copy supplier to 'supplier.parquet' (FORMAT parquet);
```

--- gcp setup

sudo apt-get install -y time g++ clang emacs git tmux mdadm

# install rust
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh

# setup drive on /dev
sudo mdadm --create --verbose /dev/md0 --level=0 --raid-devices=4 /dev/nvme1n1 /dev/nvme2n1 /dev/nvme3n1 /dev/nvme4n1
sudo mkfs -t ext4 /dev/md0
sudo mkdir /data
sudo mount /dev/md0 /data
sudo chmod -R a+rwx /data

# get tpchgen
cd /data
git clone git@github.com:alamb/tpchgen-rs.git
cd tpchgen-rs
git checkout alamb/parquet
cargo install --path tpchgen-cli

> # Note the GPC machine can only write at 756MB/s !!!!!

Reviewer: What's `GPC` ?

Author: It is supposed to be "GCP" 🤦 

```suggestion
# Note the GCP machine can only write at 756MB/s !!!!!
```

dd if=/dev/zero of=/data/test1.img bs=1G count=10 oflag=dsync
...
10737418240 bytes (11 GB, 10 GiB) copied, 13.179 s, 815 MB/s

# Install duckdb like this:
curl https://install.duckdb.org | sh

> sudo ln -s /home/alamb/.duckdb/cli/latest/duckdb /usr/local/bin

Reviewer: ```suggestion
sudo ln -s ~/.duckdb/cli/latest/duckdb /usr/local/bin
```
